src/index.ts

```diff
@@ -0,0 +1,518 @@
+   * matches, then the key will be returned.
+   *
+   * @param key   The key for looking up the string. Context should be delimited with a period.
+   * @param data  Contains data to be interpolated. `count` is a magic value for pluralization.
+   * @returns     Best string match for key given.
+   */
+  public t(key: string, data: TDataObject = {}, options: TOptions = {}): string {
+    // If we're not in production mode, be sure the people testing
+    if (process.env.NODE_ENV !== 'production') {
+      if (!key.match(VALID_KEY)) {
+        throw new Error(
+          `${key} is not a valid key format. ` +
+          `Valid keys are of the format ${VALID_KEY.toString()}`);
+      }
+    }
+
+    // If key mode is enabled, simply return the key.
+    if (this.getKeyMode()) {
+      return key;
+    }
+
+    const locales: string[] = [];
+
+    // If options has a locale, normalize it and use that instead of locale. Otherwise, check
+    // if locale is even set and add that to possible locales.
+    if (options.locale !== undefined) {
+      locales.push(this.normalizeLocale(options.locale));
+    } else if (this.locale !== undefined) {
+      locales.push(this.locale);
+    }
+
+    if (options.fallback !== undefined) {
+      locales.push(this.normalizeLocale(options.fallback));
+    } else {
+      locales.push(this.fallback);
+    }
+
+    // Loop through the locales and try to find an appropriate translation.
+    for (let i = 0; i < locales.length; i++) {
+      const locale = locales[i];
+      const contexts = key.split('.');
+
+      let translation: TranslationLookup =
+        contexts.reduce(this.translationReducer, this.translations[locale]);
+
+      // If there is a count and an object comes back, check if this key is a plural context.
+      if (typeof translation === 'object' && typeof data.count === 'number') {
+        const pluralContext: string = this.getPluralContext(locale, data.count);
+        translation = translation[pluralContext];
+      }
+
+      // If the translation is equal to string, interpolate and return.
+      if (typeof translation === 'string') {
+        return translation
+          .replace(INTERPOLATION_REGEX, (exp, arg) => {
+            const result = data[arg];
+            return typeof result === 'string' ? result :
+                   typeof result === 'number' ? `${result}` :
+                   exp;
+          });
+      }
+    }
+
+    // Fallback to the key.
+    return key;
+  }
+
+  protected callSubscribers(): void {
+    this.subscribers.map((subscriber) => subscriber());
+  }
+
+  /**
+   * Clones the translations object passed in. However, it will also verify that each context
+   * of a key is valid.
+   */
+  private cloneTranslations(translations: Translations): Translations {
+    const newTranslations = { ...translations };
+
+    Object
+      .keys(newTranslations)
+      .forEach((context) => {
```

Comment:
General taste question: do you favor `forEach` or `for ... in` ? In this case, I think I would prefer `for ... in`.

Reply:
I think if there was more chaining involved, it would be better. But, I agree with you for this case.